### ISSUE TITLE
Contribution: Opt-in Law of Demeter checker

### ISSUE THREAD

**Issue by MahatmaFatalError** [google/error-prone#5909]
# Proposal: add an opt-in Law of Demeter checker

## Summary

I would like to contribute a new built-in, opt-in Error Prone checker named `LawOfDemeter`.

The checker detects method calls on objects that are reached through another object, including simple cases that are split across local variables:

```java
void process(Order order) {
  order.getCustomer().getName(); // diagnostic

  Customer customer = order.getCustomer();
  customer.getName(); // diagnostic
}
```

The goal is not to enforce a universal style preference by default. The checker would be registered as disabled by default so projects can opt in with `-Xep:LawOfDemeter`.

## Motivation

The Law of Demeter is useful for codebases that want to limit train-wreck calls and reduce coupling to nested object graphs. A simple dot-counting check is too noisy and misses common cases once an intermediate value is assigned to a local variable. This checker instead tracks a conservative "acquaintance depth" through straight-line local assignments.

This is intended for teams that explicitly choose this architecture rule. Because the rule can be subjective and codebase-dependent, I am proposing it as an opt-in check rather than a default warning.

## Proposed behavior

The checker reports when a non-static method is invoked on a receiver whose computed acquaintance depth is greater than or equal to the configured threshold.

By default, the threshold is `1`, but configurable up to `3`.

It also catches the same pattern when the intermediate object is stored in a local variable:

```java
Customer customer = order.getCustomer();
customer.getName();
```

Direct calls on parameters, fields, locals, `this`, `super`, and freshly created objects are allowed.

## False-positive controls

The checker is intentionally conservative and includes exemptions for common fluent or wrapper-style APIs:

- fluent APIs and builders, using a same-type-return heuristic
- classes matching configurable builder-like class-name patterns
- classes annotated with configurable builder annotations
- transparent types such as streams and optionals
- configured value types such as `java.time` values and big number types
- static factory methods

The existing prototype uses these options:

- `LawOfDemeter:Threshold`
- `LawOfDemeter:ExemptClassPatterns`
- `LawOfDemeter:ExemptAnnotations`
- `LawOfDemeter:ExemptTransparentTypes`
- `LawOfDemeter:ExemptValueTypes`

## Limitations

The first version is deliberately narrow:

- It performs single-pass tracking of straight-line local assignments only.
- It does not model branch joins or loop fixpoints precisely.
- Field-access chains are not counted as multiple method-call hops.
- The fluent-call exemption is heuristic-based, so same-type-returning methods are treated as fluent even when they are not builders.

I would document these limitations in `docs/bugpattern/LawOfDemeter.md`.

## Integration plan

The implementation would live in `error_prone_core` as a normal built-in checker:

- `core/src/main/java/com/google/errorprone/bugpatterns/LawOfDemeter.java`
- `core/src/main/java/com/google/errorprone/bugpatterns/LawOfDemeterDepthCalculator.java`
- `core/src/test/java/com/google/errorprone/bugpatterns/LawOfDemeterTest.java`
- `docs/bugpattern/LawOfDemeter.md`

The checker would be registered in `BuiltInCheckerSuppliers.DISABLED_CHECKS`.

The current implementation exists as an external Error Prone plugin, but the upstream contribution would remove plugin-specific details such as `@AutoService` and ServiceLoader-oriented constructors.

## Questions for maintainers

- General feedback if I should open a PR here
- Is this kind of opt-in architecture checker appropriate for Error Prone core?
- Is `DISABLED_CHECKS` the right initial registration point?
- Are there additional exemption categories that should be included before a PR?
- Would maintainers prefer this as a single PR, or split into checker/tests and documentation/registration?

**Comment by cpovirk** [comment]
The Law of Demeter gets at some important principles, like injecting dependencies and accepting the specific value that you need, not [a larger object that contains it](https://github.com/google/error-prone/blob/65cffb75fc8cfc2a701f087b11fc7969986a6184/check_api/src/main/java/com/google/errorprone/util/ASTHelpers.java#L467-L468). But it's more of a proxy for the things that we care most about, so I don't think we'd want to take on reviewing and maintaining such a check.
